/*
  Read Analyzer

  In order to work correctly, the SAM file put into this program
  should be generated by RazerS with the following options:

    -rr 100  # 100% recognition rate
    -i 90    # 10% error rate
    -of 4    # SAM output format
    -dr 0    # Only output the best matches for each read
 */
#include <seqan/align.h>
#include <seqan/basic.h>
#include <seqan/store.h>

#include "read_analyzer.h"

using namespace seqan;

int main(int argc, char **argv) {
    // Check arguments.
    if (argc != 3) {
        std::cerr << "Usage: read_analyzer GENOME.FASTA FILE.SAM" << std::endl;
        return 1;
    }

    // =======================================================================
    // Load files.
    // =======================================================================
    typedef FragmentStore<> TFragmentStore;
    TFragmentStore fragmentStore;

    // Load contigs.
    std::cerr << "Reading FASTA contigs sequence file " << argv[1] << " ..." << std::endl;
    if (!loadContigs(fragmentStore, argv[1])) {
        std::cerr << "Could not read contigs." << std::endl;
        return 1;
    }

    // Load SAM file.
    std::cerr << "Reading SAM file file " << argv[2] << " ..." << std::endl;
    {
        std::fstream fstrm(argv[2], std::ios_base::in | std::ios_base::binary);
        if (!fstrm.is_open()) {
            std::cerr << "Could not open SAM file." << std::endl;
            return 1;
        }
        read(fstrm, fragmentStore, SAM());
    }
    std::cout << "length(fragmentStore.alignedReadStore) == " << length(fragmentStore.alignedReadStore) << std::endl;

    // =======================================================================
    // Perform evaluation.
    // =======================================================================
    std::cerr << "Evaluating..." << std::endl;
    ReadEvaluationResult readResult;
    setReadLength(readResult, length(fragmentStore.readSeqStore[0]));
    performReadEvaluation(readResult, fragmentStore);
    AlignmentEvaluationResult alignmentResult;
    setReadLength(alignmentResult, length(fragmentStore.readSeqStore[0]));
    performAlignmentEvaluation(alignmentResult, fragmentStore);

    // =======================================================================
    // Print result.
    // =======================================================================
    printReadEvaluationResults(readResult);
    printAlignmentEvaluationResults(alignmentResult, fragmentStore);
    
    return 0;
}
